x-pack/legacy/plugins/security/public/views/management/role_mappings/model/except_field_rule.ts

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { i18n } from '@kbn/i18n';
+import { RuleGroup } from './rule_group';
+import { FieldRule } from './field_rule';
+import { Rule } from './rule';
+
+/**
+ * Represents a negated field rule {@link FieldRule}.
+ */
+export class ExceptFieldRule extends RuleGroup {
+  constructor(private fieldRule: FieldRule | null = null) {
+    super();
+  }
+
+  /** {@see RuleGroup.getRules} */
+  public getRules() {
+    return this.fieldRule ? [this.fieldRule] : [];
+  }
+
+  /** {@see RuleGroup.replaceRule} */
+  public replaceRule(index: number, newRule: FieldRule) {
+    this.fieldRule = newRule;
+  }
+
+  /** {@see RuleGroup.removeRule} */
+  public removeRule() {
+    this.fieldRule = null;
+  }
+
+  /** {@see RuleGroup.addRule} */
+  public addRule(rule: FieldRule) {
+    if (this.fieldRule) {
+      throw new Error('Rule already exists. Unable to add more to "ExceptFieldRule"');
+    }
+    this.fieldRule = rule;
+  }
+
+  /** {@see RuleGroup.canAddRule} */
+  public canAddRule(): boolean {
+    return this.fieldRule === null;
+  }
+
+  /** {@see RuleGroup.canRemoveRule} */
+  public canRemoveRule(): boolean {
+    return false;
+  }
+
+  /** {@see RuleGroup.canContainRules} */
+  public canContainRules(rules: Rule[]) {
+    if (rules.length === 0) {
+      return true;
+    }
+    return this.fieldRule === null && rules.length === 1 && rules[0] instanceof FieldRule;
+  }
+
+  /** {@see RuleGroup.getDisplayTitle} */
+  public getDisplayTitle() {
+    return i18n.translate(
+      'xpack.security.management.editRoleMapping.exceptFieldRule.displayTitle',
+      {
+        defaultMessage: 'Is false',
```

Comment:
I got a bit confused when seeing this:

![image](https://user-images.githubusercontent.com/5295965/71914895-9676dd00-3148-11ea-98ca-1bd69558b5f3.png)

The two of these rules (`except_all_rule` and `except_field_rule`) are functionally equivalent. It seems that the only reason to include the `except_field_rule` is because the Elasticsearch API supports it, so we have to be able to render that type of rule.

There's already a lot going on in this UI, what do you think about potentially ditching the `except_field_rule` and just auto-converting that to an `except_all_rule` whenever the user is editing an existing role mapping?

I get that the intent of this is to provide a UI for the functionality that Elasticsearch provides, but some things just don't quite make sense when you translate them to a UI, in my opinion.

Reply:
> There's already a lot going on in this UI, what do you think about potentially ditching the except_field_rule and just auto-converting that to an except_all_rule whenever the user is editing an existing role mapping?

That's an interesting idea! I do think that'd improve the usability and readability of the rules, but I'm also hesitant to automatically convert it, mostly because we expose the raw rule definition via the JSON editor, so users who are familiar with their rule set would see that the rules they previously created don't exactly match. Even within the same editing session, going from JSON editor -> Visual editor -> JSON editor would illustrate the difference.

I do agree that the current state can get confusing though, and you're absolutely right that they both exist solely because the Elasticsearch API supports it...

Do you think the potential confusion of this rewrite is worth the ux improvement?

Reply:
Assuming this would be an easy change like it appears to be -- my personal opinion is that, if such a change was made, the total sum of confusion would be less 😁 
How many people are familiar with role mapping and would notice (or care) if we converted the rule? Probably not many.

Reply:
I tested this out: it does appear to be an easy fix and would actually simplify a couple of other code paths slightly. Let's see what others think!

Reply:
I'm also hesitant for us to automatically convert the `except_field_rule` to a `except_any_rule` because of the ability to toggle between JSON and visual editors...

Do we see any similarity to the fact that we allow users to specify the following via the JSON editor:

```
{
  "field": {
    "username": "*"
  }
}
``` 

Which we render as the following, even though the user can't create this rule via just the UI?

<img width="1207" alt="Screen Shot 2020-01-09 at 3 19 48 PM" src="https://user-images.githubusercontent.com/627123/72112778-818a7d00-32f3-11ea-8aa4-53b2fbc48137.png">

Reply:
It’s kind of similar, in that the visual editor is limiting what you’re able to do in both cases. The only difference is that the visual editor (and the persisted model) will alter what is on screen if we perform the conversion from except_field_rule. 

Joe does bring up a good point though, in that this is a bit of an edge case that most users won’t run into, let alone notice (we think)